### PR TITLE
Grade override history column filtering

### PR DESCRIPTION
**Pull request by nsprenkle**
**TL;DR:** Only show modified subsections in the Bulk Grade Override Report

**Description:** The Bulk Grade Override Report (Gradebook > Bulk Management) acts as an audit of changes introduced by a Bulk Grades Override. Currently the downloaded report shows columns for all gradable subsections, even if they were not modified as part of a Bulk Grades Override. To make the report easier to grok, this adds a filtering step which removes unmodified columns from the exported report.

The filter call has to do a lot of work to determine if rows introduce changes to subsections and remove all columns which reference the unmodified subsections in a way that is non-destructive to other columns in the report, which is why it's kind of loop-y and complex.

**JIRA:** [EDUCATOR-4655](https://openedx.atlassian.net/browse/EDUCATOR-4655)

**Linked PR:** https://github.com/edx/super-csv/pull/56 (which updates some config to allow column filtering to work) must be merged prior to merging this PR

**Merge checklist:**
- [x] All reviewers approved
- [x] CI build is green
- [x] Version bumped
- [x] Changelog record added
- [ ] Documentation updated (not only docstrings)
- [x] Commits are squashed

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPI after tag-triggered build is 
      finished.
- [ ] Delete working branch (if not needed anymore)

@edx/masters-devs-gta

### CODE REVIEW

File: bulk_grades/api.py [hunk 1]
```diff
@@ -437,6 +437,38 @@ def get_rows_to_export(self):
                     row[f'grade-{block_id}'] = effective_grade
             yield row
 
+    def filtered_column_headers(self):
+        """
+        To trim down grade exports, only show subsections which were modified in a bulk update.
+        Returns: a filtered list of columns to export, preserving modified and non-subsection columns
+        """
+        unmodified_subsections = set()
+        for subsection in self._subsections.keys():
```
Comment (jansenk) [inline bulk_grades/api.py:446]: `unmodified_subsections = set(self._subsections.keys())` ?

File: bulk_grades/api.py [hunk 2]
```diff
@@ -437,6 +437,38 @@ def get_rows_to_export(self):
...
+            unmodified_subsections.add(subsection)
+
+        for row in self.result_data:
+            # Ignore rows which didn't introduce changes
+            if row['status'] == 'No Action':
```
Comment (jansenk) [inline bulk_grades/api.py:451]: Nit: Is this string defined as a constant anywhere?

File: bulk_grades/api.py [hunk 3]
```diff
@@ -437,6 +437,38 @@ def get_rows_to_export(self):
...
+                continue
+
+            modified_by_row = set()
+
+            # Get changes this row introduced for an as-yet-unmodified subsection
+            for subsection in unmodified_subsections:
+                if row[f'new_override-{subsection}'] != '':
```
Comment (jansenk) [inline bulk_grades/api.py:458]: would this include a row / column that has just whitespace, or at this point has it been processed out?

File: bulk_grades/api.py [hunk 4]
```diff
@@ -437,6 +437,38 @@ def get_rows_to_export(self):
...
+        unmodified_subsections = set()
+        for subsection in self._subsections.keys():
+            unmodified_subsections.add(subsection)
+
+        for row in self.result_data:
+            # Ignore rows which didn't introduce changes
+            if row['status'] == 'No Action':
+                continue
+
+            modified_by_row = set()
```
Comment (jansenk) [inline bulk_grades/api.py:454]: nit: a variable named `X_by_Y` tends to reads as a dict of `X`, keyed by `Y`. Maybe just `modified_subsections` or `learner_modified_subsections`?

File: bulk_grades/api.py [hunk 5]
```diff
@@ -437,6 +437,38 @@ def get_rows_to_export(self):
...
+
+            # Get changes this row introduced for an as-yet-unmodified subsection
+            for subsection in unmodified_subsections:
+                if row[f'new_override-{subsection}'] != '':
+                    modified_by_row.add(subsection)
+
+            # Remove modified subsections from the unmodified list
+            if modified_by_row:
+                unmodified_subsections.difference_update(modified_by_row)
```
Comment (jansenk) [inline bulk_grades/api.py:463]: for some reason i just love using these set operation methods

File: tests/test_api.py [hunk 6]
```diff
@@ -530,6 +561,38 @@ def test_assignment_grade(self, mocked_graded_subsections, mocked_get_subsection
             assert learner_data_row['original_grade-lab_ques'] == '3'
             assert learner_data_row['previous_override-lab_ques'] == '5'
 
+    @patch('lms.djangoapps.grades.api.graded_subsections_for_course_id')
+    def test_filter_override_history_columns(self, mocked_graded_subsections):
```
Comment (jansenk) [inline tests/test_api.py:565]: could you toss in one for the  "zero" case, where the grades are completely unmodified and we return no subsections?

**Review by jansenk — APPROVED** [review]
some nits, no functional concerns. haven't run locally

File: bulk_grades/api.py (hunk 1 above)
Comment (nsprenkle) [reply bulk_grades/api.py:446]: Beautiful 🤩

File: bulk_grades/api.py (hunk 2 above)
Comment (nsprenkle) [reply bulk_grades/api.py:451]: No, it is set in situ in [super_csv/csv_processor.py::preprocess_file](https://github.com/edx/super-csv/blob/4eec510eb20dc07d90e8ee586cc2f86af45b602a/super_csv/csv_processor.py#L206)

File: bulk_grades/api.py (hunk 3 above)
Comment (nsprenkle) [reply bulk_grades/api.py:458]: I believe [it's been pre-processed out](https://github.com/edx/edx-bulk-grades/blob/efbc5f7b28c2c494ff7b7372823c56a101eea14a/bulk_grades/api.py#L350) or, at least, testing it manually by hitting it with a bunch of whitespace seems to function appropriately.

Comment (nsprenkle) [reply bulk_grades/api.py:458]: ^ Correction, it is not actually pre-processed out, I mistook `"No Action"`s for correctly ignoring whitespace.

File: bulk_grades/api.py (hunk 5 above)
Comment (jnlapierre) [reply bulk_grades/api.py]: I must say, they are cool.

**Review by jnlapierre — APPROVED** [review]
LGTM, also liked the Unit test additions.  Great model for testing grade reports moving forward.

File: .gitignore [hunk 7]
```diff
@@ -77,6 +78,7 @@ tests/__init__.py
 
 # Development task artifacts
 default.db
+csv/
 
 # virtualenvs
-bulk-grades-env/
\ No newline at end of file
+bulk-grades-env/
```
Comment (jansenk) [inline .gitignore:84]: this isn't a blocker but I believe the (unofficial?) standard is to make the venv the same name as the repo

File: tests/test_api.py [hunk 8]
```diff
@@ -400,7 +426,6 @@ def test_no_subsection_grade(self, mock_graded_subsections):
             row = rows[i].split(',')
             assert row[grade_column_index] == '1'
 
-
```
Comment (jansenk) [inline tests/test_api.py:403]: i don't care about this but does quality not complain?

**Review by jansenk — APPROVED** [review]
I like it! I think you just gotta bump version number

File: tests/test_api.py (hunk 8 above)
Comment (nsprenkle) [reply tests/test_api.py:403]: I don't think it enforces spacing between functions (only classes?). Either way, there was an extra newline here and I got rid of it.

File: tests/test_api.py (hunk 6 above)
Comment (nsprenkle) [reply tests/test_api.py:559]: See `test_filter_override_history_noop`
